IdentityCore/src/oauth2/ciam/MSIDCIAMOauth2Factory.m

```diff
@@ -0,0 +1,152 @@
+
++ (MSIDProviderType)providerType
+{
+    return MSIDProviderTypeCIAM;
+}
+
+#pragma mark - Helpers
+
+- (BOOL)checkResponseClass:(MSIDCIAMTokenResponse *)response
+                   context:(id<MSIDRequestContext>)context
+                     error:(NSError **)error
+{
+    if (![response isKindOfClass:[MSIDCIAMTokenResponse class]])
+    {
+        if (error)
+        {
+            NSString *errorMessage = [NSString stringWithFormat:@"Wrong token response type passed, which means wrong factory is being used (expected MSIDCIAMTokenResponse, passed %@", response.class];
+
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, errorMessage, nil, nil, nil, context.correlationId, nil, YES);
+        }
+
+        return NO;
+    }
+
+    return YES;
+}
+
+#pragma mark - Response
+
+- (MSIDCIAMTokenResponse *)tokenResponseFromJSON:(NSDictionary *)json
+                                        context:(__unused id<MSIDRequestContext>)context
+                                          error:(NSError **)error
+{
+    return [[MSIDCIAMTokenResponse alloc] initWithJSONDictionary:json error:error];
+}
+
+- (MSIDCIAMTokenResponse *)tokenResponseFromJSON:(NSDictionary *)json
+                                   refreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)token
+                                        context:(__unused id<MSIDRequestContext>)context
+                                          error:(NSError * __autoreleasing *)error
+{
+    return [[MSIDCIAMTokenResponse alloc] initWithJSONDictionary:json refreshToken:token error:error];
+}
+
+- (BOOL)verifyResponse:(MSIDCIAMTokenResponse *)response
+               context:(id<MSIDRequestContext>)context
+                 error:(NSError * __autoreleasing *)error
+{
+    if (![self checkResponseClass:response context:context error:error])
+    {
+        return NO;
+    }
+
+    return [super verifyResponse:response context:context error:error];
+}
+
+#pragma mark - Tokens
+
+- (BOOL)fillAccount:(MSIDAccount *)account
+       fromResponse:(MSIDCIAMTokenResponse *)response
+      configuration:(MSIDConfiguration *)configuration
+{
+    if (![self checkResponseClass:response context:nil error:nil])
+    {
+        return NO;
+    }
+
+    BOOL result = [super fillAccount:account fromResponse:response configuration:configuration];
+
+    if (!result)
+    {
+        return NO;
+    }
+
+    // CIAM should return preferred_username claim
+    if ([NSString msidIsStringNilOrBlank:response.idTokenObj.preferredUsername])
+    {
+        account.username = MSID_PREFERRED_USERNAME_MISSING;
+    }
+
+    return YES;
+}
+
+- (MSIDAuthority *)cacheAuthorityWithConfiguration:(MSIDConfiguration *)configuration
+                                     tokenResponse:(MSIDTokenResponse *)response
+{
+    NSError *authorityError = nil;
+    
+    MSIDAuthority *cacheAuthority = [self resultAuthorityWithConfiguration:configuration tokenResponse:response error:&authorityError];
```

Comment:
Is this different from what AAD factory is doing?

Reply:
Slight differences, keeping it in CIAMOauth2Factory to be able to see cached authority correctly when testing silent acquire token in MSAL test app (when I tried to remove this use the AAD cacheAuthorityWithConfiguration, it was not saving the user in test app, preventing me from being able to test silent token acquisition).